File: JavaScript/boj/Silver/1992.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const [N, ...input] = fs.readFileSync(filePath).toString().trim().split('\n');

const map = [];
const result = [];

for (let i = 0; i < N; i++) map[i] = input[i].split('').map(Number);

const isSame = (row, col, size) => {
  const start = map[row][col];

  for (let i = row; i < row + size; i++) {
    for (let j = col; j < col + size; j++) {
      if (start !== map[i][j]) return false;
    }
  }

  return true;
};

const divide = (row, col, size) => {
  if (!isSame(row, col, size)) {
    const divideSize = size / 2;

    result.push('(');
    divide(row, col, divideSize);
    divide(row, col + divideSize, divideSize);
    divide(row + divideSize, col, divideSize);
    divide(row + divideSize, col + divideSize, divideSize);
    result.push(')');
  } else result.push(map[row][col]);
};

divide(0, 0, N);
console.log(result.join(''));

File: Python/boj/Silver/1992.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    map = [list(map(int, input())) for _ in range(N)]
    result = []

    def isSame(row, col, size):
        start = map[row][col]

        return all(
            start == map[i][j]
            for i in range(row, row + size)
            for j in range(col, col + size)
        )

    def divide(row, col, size):
        if not isSame(row, col, size):
            divideSize = size // 2

            result.append("(")
            divide(row, col, divideSize)
            divide(row, col + divideSize, divideSize)
            divide(row + divideSize, col, divideSize)
            divide(row + divideSize, col + divideSize, divideSize)
            result.append(")")
        else:
            result.append(str(map[row][col]))

    divide(0, 0, N)
    print("".join(result))
